### ISSUE TITLE
Git submodules update to latest tag?

### ISSUE THREAD

**Issue by hazcod** [dependabot/dependabot-core#1639]
Hi, is it possible to make dependabot only update the git submodules to the next tag and not the latest commit?

```
version: 1

update_configs:
- package_manager: "submodules"
  directory: .
  update_schedule: "daily"
```

**Comment by feelepxyz** [comment]
@hazcod 👋 unfortunately not, the update checking logic for submodules is pretty basic atm: https://github.com/dependabot/dependabot-core/blob/master/git_submodules/lib/dependabot/git_submodules/update_checker.rb#L48

Not entirely sure how we could add this functionality without adding submodule specific configuration.

**Comment by hazcod** [comment]
Hi @feelepxyz !
What about checking whether the submodule is at a tag? If it is, you can update to the next tag.
If not, just use the latest commit.

In bash:
```bash
niels@BEDGBWN10I20004:/mnt/c/Users/nihofman/Downloads/iron-argo$ git describe --tags
fatal: No tags can describe '91ac5a59a5f41a0e99ab631b24e7f1544a1e5967'.
$ echo $?
128
$ git tag -l
2020.01.21
$ git checkout 2020.01.21
HEAD is now at a04cc38 Add deploy CI/CD + Fix scanning
$ git describe --tags
2020.01.21
$ echo $?
0
```

**Comment by feelepxyz** [comment]
@hazcod yes good shout, that would make sense! We're pretty stretched at the moment so won't be able to get to this anytime soon. We would be very happy to review any contributions if you're super keen to add support in dependabot-core.

**Comment by scouten** [comment]
Adding my :+1: to this issue.

**Comment by landfillbaby** [comment]
idea: if a module in `.gitmodules` has a `dependabot-tags-only = true` line it only sends pull requests for tags

**Comment by deivid-rodriguez** [comment]
@landfillbaby I think the suggestion by @hazcod at https://github.com/dependabot/dependabot-core/issues/1639#issuecomment-576721613 is a good idea. Does it not satisfy your use case?

**Comment by landfillbaby** [comment]
i just wrote an [update-submodules.yml](https://github.com/landfillbaby/png2webp/blob/main/.github/workflows/update-submodules.yml)

**Comment by deivid-rodriguez** [comment]
@landfillbaby Is that the same strategy that was suggested above? If yes, feedback on how that works in practice appreciated :)

Also, I think #2192 a duplicate of this, correct?

**Comment by landfillbaby** [comment]
it works well :) i'm specifically using the latest tag on the head branch, since libpng has beta tags
and #2192 is slightly different since the want specifically github releases rather than just tags

**Comment by deivid-rodriguez** [comment]
But Github releases are always backed with a git tag, so I think these should be mostly equivalent? I guess this way could create unexpected updates if tags are pushed that have no corresponding release, but still would be a nice improvement over the current situation and coupling to GitHub releases might not be necessary for most.

**Comment by ReenigneArcher** [comment]
+1 on this...

I would be happy if the PR description just said something like "tagged commit: true/false"

**Comment by dabico** [comment]
Any updates on this feature? I have a [project](https://github.com/seart-group/java-tree-sitter) with almost 50 submodules, most of which require being anchored to a specific tagged version. As you can imagine, updating manually is a pain, while pushing updates to the latest commit may break builds.

**Comment by deivid-rodriguez** [comment]
No, no updates. Someone need to work on this. We're happy to take PRs!

**Comment by ReenigneArcher** [comment]
@dabico You may want to investigate renovate bot. It's a bit more customizable than dependabot.

**Comment by etan-status** [comment]
Have implemented it in #13052 , feel free to provide feedback.

**Comment by Peter-Darton-i2** [comment]
Having just "inherited" a legacy codebase, I've had to manually update git submodules recently because dependabot was silently failing to do anything with them.
I was "somewhat surprised" to find that *github* dependabot did not support *github* releases like this (it seems to support everyone else's stuff but not its own); I guess my codebase's original owners assumed dependabot supported this too, as they wrote a config file for it.

My own usecase, I've got some git submodules (e.g. pointing at https://github.com/bats-core/bats-assert) and every time there's a new (later) release, I want dependabot to give it to me ... but I don't want any non-tagged commits.
e.g. if my repo has bats-assert at git tag `v2.1.0` and tag `v2.2.0` arrives then I want dependabot to make a PR for me ... but not if there's just a new commit on the main branch that _isn't_ tagged as a release.

TL;DR: I applaud your efforts and will give things a try once it's "live".

**Comment by etan-status** [comment]
Yeah, that's what #13052 does. It prefers commits with a semver tag (with optional v prefix) over untagged ones. Unless you already are on an untagged one, then it will track the latest commit until encountering the next tag (after which it will resume tracking the tag once again).

**Comment by v-thavaahariharangit** [comment]
Closing this with this PR merged https://github.com/dependabot/dependabot-core/pull/13052

**Comment by perrin4869** [comment]
is there a way to opt-out of this behavior across a repo? my [dotfiles](https://github.com/perrin4869/dotfiles/blob/617b1da88b969cf846a3404c9efbd72ec0074dce/.gitmodules#L1) have over 100 submodules and I would like them to keep tracking the branch head commit, but with this change, if any module ever lands on a tagged commit, even if by coincidence, if understand correctly, it will only track tagged commits from that point forward?

**Comment by v-thavaahariharangit** [comment]
> is there a way to opt-out of this behavior across a repo? my [dotfiles](https://github.com/perrin4869/dotfiles/blob/617b1da88b969cf846a3404c9efbd72ec0074dce/.gitmodules#L1) have over 100 submodules and I would like them to keep tracking the branch head commit, but with this change, if any module ever lands on a tagged commit, even if by coincidence, if understand correctly, it will only track tagged commits from that point forward?

@etan-status ^^^

**Comment by etan-status** [comment]
> if any module ever lands on a tagged commit, even if by coincidence, if understand correctly, it will only track tagged commits from that point forward?

Only if the tag follows semver convention (v1.2.3 format), other tags don't affect this.

@thavaahariharangit I can take a look if there's a clean way to specify track-latest-head explicitly, so that this edge case can be satisfied. Having a way to _ignore_ versions seems reasonable to me.

**Comment by etan-status** [comment]
@perrin4869 @thavaahariharangit In dependabot-core, internally a 0.0.0 pseudo-version is assigned to each commit, the commits with semver tags get additional non-0.0.0 entries _on top_ of the pseudo-version (an additional one for each tag, in case there are multiple). You can use `ignore` to filter out the tags and only keep the 0.0.0 pseudo-versions, effectively restoring the previous behaviour where the latest HEAD is tracked:

```yaml
updates:
  - package-ecosystem: "gitsubmodule"
    directory: "/"
    schedule:
      interval: "daily"
    ignore:
      - dependency-name: "*"
        versions: ["> 0.0.0"]
```

There is currently a bug in Dependabot that makes `ignore` not work properly for gitmodules, I have raised a PR to fix that bug and also added tests to cover the opt-out specifically:

- https://github.com/dependabot/dependabot-core/pull/14352

**Comment by perrin4869** [comment]
Thank you! I will set this option on my repo then, thanks for the workaround!

**Comment by ReenigneArcher** [comment]
Looks like this is live now, awesome!

One enhancement idea would be to use the tags in the PR title instead of the short commit hash.

<img width="948" height="97" alt="Image" src="https://github.com/user-attachments/assets/12ca5211-2e94-4066-85fb-e3f72d90d0c1" />

**Comment by Frulfump** [comment]
> Looks like this is live now, awesome!
> 
> One enhancement idea would be to use the tags in the PR title instead of the short commit hash.
> 
> <img width="948" height="97" alt="Image" src="https://github.com/user-attachments/assets/12ca5211-2e94-4066-85fb-e3f72d90d0c1" />

Good idea, could you open a new issue for that?
